packages/flutter/lib/src/widgets/sheet_route.dart

```diff
@@ -0,0 +1,266 @@
+import 'dart:ui';
+import 'package:flutter/foundation.dart';
+import 'package:flutter/widgets.dart';
+
+import 'sheet.dart';
+
+/// Route that defines the transition animation of the previous route when this
+/// one is closing
+mixin RouteWithPreviousTransitionMixin<T> on Route<T> {
+  Widget getPreviousRouteTransition(
+    BuildContext context,
+    Animation<double> secondAnimation,
+    Widget child,
+  );
+}
+
```

Comment:
I think this would be a good way to start. I could make a different PR with this code to allow back transitions to be defined from the next Route.

Right now it consists of two mixins: 
`AllowsRouteWithPreviousTransitionMixin` -> Assigned to the route that will transition with the custom animation eg. MaterialPageRoute
`RouteWithPreviousTransitionMixin` -> Assigned to the route that will be pushed and will define the transition for the previous route eg. CupertinoSheetRoute.

If you agree we can move this to another PR and I will start working on the tests and docs.